### PR TITLE
ARROW-15888: [Doc][Python] Modernize development instructions

### PR DESCRIPTION
**Pull request by pitrou**
Some Python development instructions are slightly outdated, update them.

Also fix a couple issues that raised some Sphinx warnings.

### CODE REVIEW

**Review by xhochy — APPROVED** [review]

File: docs/source/developers/python.rst [hunk 1]
```diff
@@ -146,15 +146,18 @@ Pull in the test data and setup the environment variables:
...
-   git submodule update
+   git submodule update --init
    export PARQUET_TEST_DATA="${PWD}/cpp/submodules/parquet-testing/data"
    export ARROW_TEST_DATA="${PWD}/testing/data"
    popd
 
 Using Conda
 ~~~~~~~~~~~
 
+The `conda <https://conda.io/>`_ package managers allows installing build-time
```
Comment (jorisvandenbossche) [inline docs/source/developers/python.rst:157]:
```suggestion
The `conda <https://conda.io/>`_ package manager allows installing build-time
```

File: docs/source/developers/python.rst [hunk 2]
```diff
@@ -378,20 +378,8 @@ Docker examples
...
-complete build and test from source both with the conda and pip/virtualenv
-build methods.
-
-Building with CUDA support
-~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-The :mod:`pyarrow.cuda` module offers support for using Arrow platform
-components with Nvidia's CUDA-enabled GPU devices. To build with this support,
-pass ``-DARROW_CUDA=ON`` when building the C++ libraries, and set the following
-environment variable when building pyarrow:
```
Comment (jorisvandenbossche) [inline docs/source/developers/python.rst:390]: This section is also included elsewhere?

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

File: docs/source/developers/python.rst (hunk 2 above)
Comment (pitrou) [reply docs/source/developers/python.rst:390]: It's simply covered by the bulleted list of optional components above. There is no need for a CUDA-specific paragraph IMHO.

**Pull request closed without merging**